<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8" />
<title>例外処理 | hydrocul のメモ</title>
<link rel="stylesheet" href="https://raw.github.com/necolas/normalize.css/master/normalize.css" type="text/css" />
<link rel="stylesheet" href="../common/css/page.css" type="text/css" />
<script src="//ajax.googleapis.com/ajax/libs/jquery/1.8.1/jquery.min.js"></script>
<script src="../common/js/page.js"></script>
</head>
<body>

  <div id="breadcrumb">
    <a href="../">Home</a> &gt; <a href="../programming_languages_diff/">プログラミング言語の比較</a> &gt; 例外処理
  </div>
  <div id="side-menu">
    <div id="side-menu-attention">
      このサイトは筆者(hydrocul)の個人メモの集合です。
    </div>
  </div>

<div id="content">
  <h1 id="section">例外処理</h1>

<h2 id="section-1">書き方</h2>

<h3 id="java-20130625">Java <span class="article-date">2013/06/25</span></h3>

<pre><code>try {
  ...
  throw ...
  ...
} catch (FileNotFoundException e) {
  ...
} catch (IOException e) {
  ...
} finally {
  ...
}
</code></pre>

<p>Java7から使える try-with-resources 構文では以下のように書ける。finally節を書かなくても try のカッコの中で生成したオブジェクトを確実にcloseしてくれる。</p>

<pre><code>try (FileInputStream in = new FileInputStream(...)) {
  ...
} catch (FileNotFoundException e) {
  ...
} catch (IOException e) {
  ...
}
</code></pre>

<h3 id="scala-20130705">Scala <span class="article-date">2013/07/05</span></h3>

<pre><code>try {
  ...
  throw ...
  ...
} catch {
  case e: IOException =&gt; ...
  case e: FileNotFoundException =&gt; ...
  case e =&gt; ...
} finally {
  ...
}
</code></pre>

<h3 id="ruby-20130804">Ruby <span class="article-date">2013/08/04</span></h3>

<pre><code>begin
  ...
  raise ...
  ...
rescue
  ...
else
  ...
ensure
  ...
end
</code></pre>

<p><code>rescue</code> で例外オブジェクトを利用したい場合は、以下のように書く。</p>

<pre><code>rescue =&gt; e
</code></pre>

<p>エラーメッセージは、 <code>e.message</code> で取得できる。</p>

<p><code>rescue</code> で特定のクラスの例外を補足したい場合は、以下のように書く。</p>

<pre><code>rescue ZeroDivisionError =&gt; e
</code></pre>

<p><code>rescue</code> 節は複数書くことができる。</p>

<p>例外をスローするには、以下のような書き方がある。</p>

<pre><code>raise "exception message ..."
raise ZeroDivisionError, "message ..."
</code></pre>

<h3 id="python-20130417">Python <span class="article-date">2013/04/17</span></h3>

<pre><code>try:
  ...
  raise "message ..."
  raise HogeException()
  ...
except:
  ...
finally:
  ...
</code></pre>

<p><code>except</code> で例外オブジェクトを扱う場合は、 <code>except ex:</code> のように書く。</p>

<p>特定の例外クラスだけ <code>except</code> で扱う場合は、 <code>except HogeException:</code>
または <code>except HogeException, ex:</code> のように書く。</p>

<p><code>except</code> の中で同じ例外を再スローするには <code>raise</code> とだけ書く。</p>

<h3 id="javascript-20130416">JavaScript <span class="article-date">2013/04/16</span></h3>

<pre><code>try {
  ...
  throw "message ...";
  ...
} catch(ex){
  ...
} finally {
  ...
}
</code></pre>

<h3 id="coffeescript-20130416">CoffeeScript <span class="article-date">2013/04/16</span></h3>

<pre><code>try
  ...
  throw("message")
  ...
catch ex
  ...
finally
  ...
</code></pre>

<h2 id="section-2">チェック例外と非チェック例外</h2>

<h3 id="java-20130718">Java <span class="article-date">2013/07/18</span></h3>

<p>チェック例外をスローする可能性のあるメソッドには、メソッドシグニチャにthrowsが必要。チェック例外は Exception でかつ RuntimeException でないもの。非チェック例外は Error または RuntimeException。</p>

<h3 id="scala-20130718">Scala <span class="article-date">2013/07/18</span></h3>

<p>Javaのチェック例外も含めすべての例外が非チェック例外の扱いになる。</p>


</div>

  <footer>
    <a href="../">目次</a>
    <!-- Copyright &copy; 2012- @hydrocul All Rights Reserved. -->
  </footer>

</body>
</html>
